### DISCUSSION TITLE
Is there a way to "transform" an event before being emitted to a member of a room?

### DISCUSSION THREAD

**Discussion by mikecann** [socketio/socket.io#4050] [Q&A] [answered]
As an example of what I mean:

Say we have a user with the type:

```ts
interface User {
   id: string;
   name: string;
   bankBalance: number;
}
```

Now something happens on the server that changes the user's name. We update the database and emit the event "updated" to the "user:abc" room with the payload `{ id: "abc", name: "newname", balance: 123456 }`

Now say we have connectionA and connectionB. Both are joined to the "user:abc" room.

What I want is connectionA to see the full payload (because of their admin role for example) but I want connectionB ONLY to see a restricted version of the payload:  `{ id: "abc", name: "newname" }`. 

Is this possible? Im imagining some sort of "transformer" or "middleware" that can intercept the event before its sent along the socket.

**Comment by darrachequesne** [comment, marked as answer]
I think the easiest way to achieve this is to use two rooms, one for the full payload and one for the restricted one:

```js
// full payload
io.to("user:abc:admin").emit("updated", { id: "abc", name: "newname", balance: 123456 });

// restricted payload
io.to("user:abc").emit("updated", { id: "abc", name: "newname" });
```

Reference: https://socket.io/docs/v4/rooms/

This can also be achieved with a namespace:

```js
io.of("/admin").to("user:abc").emit("updated", { id: "abc", name: "newname", balance: 123456 });
```

Reference: https://socket.io/docs/v4/namespaces/

**Reply by mikecann** [reply]
Oh cool thanks :)
